scylla/src/transport/session.rs

```diff
@@ -860,6 +862,13 @@ impl Session {
         self.metrics.clone()
     }
 
+    /// Access keyspaces details collected by the driver
+    /// Driver collects various schema details like tables, partitioners, columns, types.
+    /// They can be read using this method
+    pub fn get_keyspaces(&self) -> HashMap<String, Keyspace> {
```

Comment:
From what I know, `HashMap<String, Keyspace>` can be quite heavy, as it contains description of all tables and types. Cloning it on each `get_keyspace` call is expensive, so it's worth getting rid of it. Returning a raw reference is not an option because of lifetime issues, but it could be worked around by returning an `Arc<Hashmap<_,_>>`, or a struct implementing `Deref` (similar to `MutexGuard`):
```rust
pub fn get_keyspaces(&self) -> KeyspacesRef {
    KeyspacesRef { cluster_data: self.cluster.get_data() }
}
```
```rust
#[derive(Clone)]
pub struct KeyspacesRef {
    cluster_data: Arc<ClusterData>
}

impl Deref for KeyspacesRef {
    type Target = HashMap<String, Keyspace>;

    fn deref(&self) -> &Self::Target {
        &self.cluster_data.keyspaces
    }
}
```

It could be used in a normal way:

```rust
let keyspaces = session.get_keyspaces();
println!("keyspace system = {:?}", keyspaces.get("system").unwrap());
```

The advantage of returning `KeyspaceRef` over `Arc<Hashmap<_,_>>` is that you do not have to create an `Arc` in the middle of `ClusterData`.

Reply:
Isn't it weird that `KeyspacesRef` actually points to `ClustedData` not `Keyspace`?

Reply:
Weirdness is subjective ;p
It has to point to `ClusterData`, because `ClusterData` owns `Keyspace`, and wee need to keep the owner alive while having a reference to its field - `Keyspace`. Besides, `KeyspaceRef` fields are invisible to the library user, so it won't be confusing (I hope).

@piodul any thoughts about this?

Reply:
@havaker @Ponewor Sometimes such a pattern is occasionally useful (e.g. [`MappedMutexGuard`](https://docs.rs/tokio/1.14.0/tokio/sync/struct.MappedMutexGuard.html) in tokio) but I don't think it is necessary here - we can just return `Arc<ClusterData>` directly. Additionally, please make its fields `pub(crate)` instead of `pub`, and provide a `pub` getter which returns a reference to the hashmap. We will probably want to expose more of its fields in the future, but for now let's restrict ourselves to the per-keyspace metadata.

Reply:
I don't get it? What should be returning `Arc<ClusterData>`?

Reply:
The idea is to write two getters:

- `Session::get_cluster_data(&self) -> Arc<ClusterData>`
- `ClusterData::get_keyspace_info(&self) -> &HashMap<String, Keyspace>`